langchain-core/src/language_models/chat_models.ts

```diff
@@ -293,6 +296,104 @@ export abstract class BaseChatModel<
     return output;
   }
 
+  /** @ignore */
+  async _generateCached(
+    messages: BaseMessageLike[][],
+    cache: BaseCache<Generation[]>,
+    parsedOptions: this["ParsedCallOptions"],
+    handledOptions: RunnableConfig
+  ): Promise<LLMResult & { missingPromptIndices: number[] }> {
+    const baseMessages = messages.map((messageList) =>
+      messageList.map(coerceMessageLikeToMessage)
+    );
+
+    // create callback manager and start run
+    const callbackManager_ = await CallbackManager.configure(
+      handledOptions.callbacks,
+      this.callbacks,
+      handledOptions.tags,
+      this.tags,
+      handledOptions.metadata,
+      this.metadata,
+      { verbose: this.verbose }
+    );
+    const extra = {
+      options: parsedOptions,
+      invocation_params: this?.invocationParams(parsedOptions),
+      batch_size: 1,
+      cached: true,
+    };
+    const runManagers = await callbackManager_?.handleChatModelStart(
+      this.toJSON(),
+      baseMessages,
+      undefined,
+      undefined,
+      extra,
+      undefined,
+      undefined,
+      handledOptions.runName
+    );
+
+    // generate results
+    const llmStringKey =
+      this._getSerializedCacheKeyParametersForCall(parsedOptions);
+    const missingPromptIndices: number[] = [];
+    const results = await Promise.allSettled(
+      baseMessages.map(async (baseMessage, index) => {
+        // Join all content into one string for the prompt index
+        const prompt =
+          BaseChatModel._convertInputToPromptValue(baseMessage).toString();
+        const result = await cache.lookup(prompt, llmStringKey);
+
+        if (result == null) {
+          missingPromptIndices.push(index);
+        }
+
+        return result;
+      })
+    );
+
+    // Ignore results with a null value. These are just absent from the cache.
+    const cachedResults = results.filter(
+      (result) =>
+        (result.status === "fulfilled" && result.value != null) ||
+        result.status === "rejected"
+    ) as PromiseSettledResult<Generation[]>[];
```

Comment:
Do we need this? Can you just do:

```
const cachedResults: PromiseSettledResult<ChatGeneration[]>[] = results.filter(
```

Reply:
Actually looking at this more, if something gets filtered out, it would result in a call to the improper callback handler below? Since we're not filtering out `runManagers`

I think avoiding the filter and just returning `results` directly from this function after calling all the callback handlers is neater. Then we can handle missing prompt indices outside of this function.

Reply:
Sorry for the super late response. Great catch! I hadn't noticed this bug. I'm still inclined to keeping the callback handler functionality within `_generateCached` since that's the bulk of what the function is really doing and abstracting away from the `generate` function that calls it. 

I did come up with a workaround though by mapping together the `runManagers` with the cached results before filtering. Then, rather than indexing into `runManagers` while calling the callback managers, we simply call the mapped `runManager`.